core/src/test/scala/akka/kafka/scaladsl/IntegrationSpec.scala

```diff
@@ -418,5 +420,109 @@ class IntegrationSpec extends TestKit(ActorSystem("IntegrationSpec"))
 
       }
     }
+
+    "successful consume transform produce transaction" in {
+      assertAllStagesStopped {
+        val sourceTopic = createTopic(1)
+        val sinkTopic = createTopic(2)
+        val group = createGroup(1)
+
+        givenInitializedTopic(sourceTopic)
+        givenInitializedTopic(sinkTopic)
+
+        Await.result(produce(sourceTopic, 1 to 100), remainingOrDefault)
+
+        val consumerSettings = createConsumerSettings(group)
+
+        val control = Consumer.transactionalSource(consumerSettings, TopicSubscription(Set(sourceTopic)))
+          .filterNot(_.record.value() == InitialMsg)
+          .map { msg =>
+            ProducerMessage.Message(
+              new ProducerRecord[Array[Byte], String](sinkTopic, msg.record.value), msg.partitionOffset)
+          }
+          .via(Producer.transactionalFlow(producerSettings, group))
+          .toMat(Sink.ignore)(Keep.left)
+          .run()
+
+        val probeConsumerGroup = createGroup(2)
+        val probeConsumerSettings = createConsumerSettings(probeConsumerGroup)
+          .withProperties(ConsumerConfig.ISOLATION_LEVEL_CONFIG -> "read_committed")
+
+        val probeConsumer = Consumer.plainSource(probeConsumerSettings, TopicSubscription(Set(sinkTopic)))
+          .filterNot(_.value == InitialMsg)
+          .map(_.value())
+          .runWith(TestSink.probe)
+
+        probeConsumer
+          .request(100)
+          .expectNextN((1 to 100).map(_.toString))
+
+        probeConsumer.cancel()
+        Await.result(control.shutdown(), remainingOrDefault)
+      }
+    }
+
+    "successful consume transform produce transaction with transient failure causing an abort with restartable source" in {
+      assertAllStagesStopped {
+        val sourceTopic = createTopic(1)
+        val sinkTopic = createTopic(2)
+        val group = createGroup(1)
+
+        givenInitializedTopic(sourceTopic)
+        givenInitializedTopic(sinkTopic)
+
+        Await.result(produce(sourceTopic, 1 to 1000), remainingOrDefault)
+
+        val consumerSettings = createConsumerSettings(group)
+
+        var restartCount = 0
+        var innerControl = null.asInstanceOf[Control]
+
+        val restartSource = RestartSource.onFailuresWithBackoff(
+          minBackoff = 0.1.seconds,
+          maxBackoff = 1.seconds,
+          randomFactor = 0.2
+        ) { () =>
+          restartCount += 1
+          Consumer.transactionalSource(consumerSettings, TopicSubscription(Set(sourceTopic)))
+            .filterNot(_.record.value() == InitialMsg)
+            .map { msg =>
+              if (msg.record.value().equals("500") && restartCount < 2) {
```

Comment:
the `producer` above looks like it produces integers (`1 to 1000`) but this check is against a `String` value. (maybe there's some .toString transformation in between, but it's not obvious from the test)

Reply:
Yes, the existing integration test infrastructure uses key, value serdes of `Array[Byte]`, `String`.